File: apps/kpai/mobile/lib/url_validator.dart
/// Default public URL baked in at build time.
/// Override at build with: --dart-define=KPAI_PUBLIC_URL=https://kidplayai.techseeding.com.au
const String kpaiPublicUrl = String.fromEnvironment(
  'KPAI_PUBLIC_URL',
  defaultValue: 'http://localhost:9512',
);

final RegExp _previewPath = RegExp(r'^/api/sandbox/[^/]+/preview/?$');

/// Returns the canonicalized URL if [raw] is a valid sandbox preview URL
/// matching the configured public origin, otherwise null.
String? validateSandboxPreviewUrl(String? raw) {
  if (raw == null || raw.trim().isEmpty) return null;
  final Uri? uri = Uri.tryParse(raw.trim());
  if (uri == null || !uri.hasScheme || !uri.hasAuthority) return null;

  final Uri base = Uri.parse(kpaiPublicUrl);
  if (uri.scheme != base.scheme) return null;
  if (uri.host != base.host) return null;
  if (uri.port != base.port) return null;

  if (!_previewPath.hasMatch(uri.path)) return null;
  return uri.toString();
}

File: apps/kpai/mobile/ios/Sources/URLValidator.swift
import Foundation

enum URLValidator {
    // The expected base URL is read from Info.plist (key: KPAIPublicURL) so the
    // app can be repointed at a dev/staging server by editing the plist or via
    // a build setting, without touching source.
    static let publicURL: URL = {
        if let raw = Bundle.main.object(forInfoDictionaryKey: "KPAIPublicURL") as? String,
           let url = URL(string: raw) {
            return url
        }
        return URL(string: "https://kidplayai.techseeding.com.au")!
    }()

    /// Returns the URL if `raw` matches `<KPAIPublicURL>/api/sandbox/<id>/preview`,
    /// otherwise `nil`. Scheme + host + port must all match the configured base.
    static func validate(_ raw: String) -> URL? {
        guard let url = URL(string: raw.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            return nil
        }
        guard url.scheme == publicURL.scheme,
              url.host == publicURL.host,
              url.port == publicURL.port else {
            return nil
        }
        let parts = url.path.split(separator: "/").map(String.init)
        guard parts.count == 4,
              parts[0] == "api",
              parts[1] == "sandbox",
              !parts[2].isEmpty,
              parts[3] == "preview" else {
            return nil
        }
        return url
    }
}
